diff --git a/src/Makefile b/src/Makefile
index 8ad0e2c..d41d63c 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,16 +1,16 @@
-OCAMLC=ocamlc
-OCAMLOPT=ocamlopt
-OCAMLDEP=ocamldep
-OCAMLMKLIB=ocamlmklib
-OCAMLFIND=ocamlfind
+OCAMLC=$(shell which ocamlc)
+OCAMLOPT=$(shell which ocamlopt)
+OCAMLDEP=$(shell which ocamldep)
+OCAMLMKLIB=$(shell which ocamlmklib)
+OCAMLFIND=$(shell which ocamlfind)
 INSTALL_DATA=install -m 644
 INSTALL_DLL=install
 INSTALL_DIR=install -d
 STDLIBDIR=$(shell $(OCAMLC) -where)
 DESTDIR ?=
 
-include $(STDLIBDIR)/Makefile.config
 
+include $(STDLIBDIR)/Makefile.config
 ifeq "$(filter i386 amd64 arm64 power,$(ARCH))" ""
 # Unsupported architecture
 BNG_ARCH=generic
@@ -86,14 +86,14 @@ endif
 VERSION=$(shell sed -ne 's/^ *version *: *"\([^"]*\)".*$$/\1/p' ../num.opam)
 
 install:
-	$(INSTALL_DIR) $(DESTDIR)$(STDLIBDIR)
+	$(INSTALL_DIR) $(LIBDIR)
 	sed -e 's/%%VERSION%%/$(VERSION)/g' META.in > META
 	$(OCAMLFIND) install num META
 	rm -f META
-	$(INSTALL_DATA) $(TOINSTALL) $(DESTDIR)$(STDLIBDIR)
+	$(INSTALL_DATA) $(TOINSTALL) $(LIBDIR)
 ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
-	$(INSTALL_DIR) $(DESTDIR)$(STDLIBDIR)/stublibs
-	$(INSTALL_DLL) $(TOINSTALL_STUBS) $(DESTDIR)$(STDLIBDIR)/stublibs
+	$(INSTALL_DIR) $(LIBDIR)/stublibs
+	$(INSTALL_DLL) $(TOINSTALL_STUBS) $(LIBDIR)/stublibs
 endif
 
 findlib-install:
@@ -105,9 +105,9 @@ findlib-uninstall:
 	$(OCAMLFIND) remove num
 
 uninstall: findlib-uninstall
-	cd $(DESTDIR)$(STDLIBDIR) && rm -f $(TOINSTALL)
+	cd $(LIBDIR) && rm -f $(TOINSTALL)
 ifeq "$(SUPPORTS_SHARED_LIBRARIES)" "true"
-	cd $(DESTDIR)$(STDLIBDIR)/stublibs && rm -f $(TOINSTALL_STUBS)
+	cd $(LIBDIR)/stublibs && rm -f $(TOINSTALL_STUBS)
 endif
 
 clean:
