From dc799fd2a66c41ca7729201a5d038cd403ca1de6 Mon Sep 17 00:00:00 2001
From: Gabriel Scherer <gabriel.scherer@gmail.com>
Date: Tue, 27 Mar 2018 12:00:23 +0200
Subject: [PATCH 2/4] add ppx_sexp_conv as a runtime dependency in the
 packaging metadata

---
 opam     | 2 +-
 pkg/META | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/opam b/opam
index ad1dbc7..c35570b 100644
--- a/opam
+++ b/opam
@@ -20,7 +20,7 @@ depends: [
   "topkg" {build}
   "cpuid" {build}
   "ocb-stubblr" {build}
-  "ppx_sexp_conv" {build}
+  "ppx_sexp_conv"
   "oUnit" {test}
   "cstruct"
   "zarith"
diff --git a/pkg/META b/pkg/META
index 242b2bb..a7929c7 100644
--- a/pkg/META
+++ b/pkg/META
@@ -1,6 +1,6 @@
 version = "0.5.4"
 description = "Simple crypto for the modern age"
-requires = "cstruct zarith sexplib"
+requires = "cstruct zarith sexplib ppx_sexp_conv"
 archive(byte) = "nocrypto.cma"
 archive(native) = "nocrypto.cmxa"
 plugin(byte) = "nocrypto.cma"
-- 
2.18.0

